### DISCUSSION TITLE
Bokeh and anywidget

### DISCUSSION THREAD

**Discussion by peekxc** [bokeh/bokeh#14200] [Internals and design (Q&A)]
Is there any advice on how to develop a custom extension w/ Bokeh in ways that are compatible with anywidget? 

I'm trying to build custom extension using Bokeh's extension API, but it's slow, difficult, and error-prone to do due to a things like the (large) typescript source and the lack of a [progressive dev workflow](https://anywidget.dev/en/getting-started/#progressive-development). I also just don't know Bokeh's TS source well enough to make certain decisions which lie very far outside of whats documented (when is is appropriate to extend LayoutDOMView vs LayoutDOM? What are the interactions between a given model and bokeh figures? How do I handle callback events like selections?)

What's the workflow like for the core Bokeh developers? What tools do you use to e.g. debug Python / JS interactions and verify functionality?

**Comment by Shinu-Cherian** [comment]
Hi @peekxc, these are really good questions. It can definitely be a bit tricky to get used to the Bokeh TS codebase at first.

Regarding anywidget compatibility, they mostly operate in different spaces since anywidget focuses on Jupyter comms and standard web components. You could technically mount an anywidget inside a Bokeh view by managing the DOM manually, but it's usually much easier to use Panel for this. Panel is built on Bokeh and has native support for bi-directional communication and modern ESM components.

For LayoutDOM vs LayoutDOMView:
LayoutDOM is just the data model that mirrors your Python class. It holds the state but doesn't know anything about the browser DOM. LayoutDOMView is what actually renders the HTML, handles the canvas, and attaches event listeners. Basically, data goes in the model, and DOM manipulation goes in the view.

For handling callbacks between Python and JS:
To send data from Python to JS, you can listen to property changes in your view's connect_signals method using `this.connect(this.model.properties.my_value.change, () => { ... })`. 
To send data from JS back to Python, you just update the model property directly from the view like `this.model.my_value = "new state"`, and Bokeh handles syncing it back automatically.

As for the workflow, if compiling TS is slowing you down, you can write inline pure JS extensions using __implementation__ in Python for quick prototyping. For debugging, inserting console.log or debugger and checking the live objects in Chrome DevTools is usually the best approach. 

If you want a smoother dev experience with hot-reloading without dealing with the TS boilerplate, checking out Panel's custom components might save you a lot of time. Hope this helps a bit!
